src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostOnAutoInsertEndpoint.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Protocol.AutoInsert;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+using Microsoft.VisualStudio.Razor.LanguageClient.Extensions;
+
+using RazorLSPConstants = Microsoft.VisualStudio.Razor.LanguageClient.RazorLSPConstants;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.LanguageClient.Cohost;
+
+#pragma warning disable RS0030 // Do not use banned APIs
+[Shared]
+[CohostEndpoint(VSInternalMethods.OnAutoInsertName)]
+[Export(typeof(IDynamicRegistrationProvider))]
+[ExportCohostStatelessLspService(typeof(CohostOnAutoInsertEndpoint))]
+[method: ImportingConstructor]
+#pragma warning restore RS0030 // Do not use banned APIs
+internal class CohostOnAutoInsertEndpoint(
+    IRemoteServiceProvider remoteServiceProvider,
+    IHtmlDocumentSynchronizer htmlDocumentSynchronizer,
+    LSPRequestInvoker requestInvoker,
+    IRazorDocumentMappingService razorDocumentMappingService,
+    ILoggerFactory loggerFactory)
+    : AbstractRazorCohostDocumentRequestHandler<VSInternalDocumentOnAutoInsertParams, VSInternalDocumentOnAutoInsertResponseItem?>, IDynamicRegistrationProvider
+{
+    private readonly IRemoteServiceProvider _remoteServiceProvider = remoteServiceProvider;
+    private readonly IHtmlDocumentSynchronizer _htmlDocumentSynchronizer = htmlDocumentSynchronizer;
+    private readonly LSPRequestInvoker _requestInvoker = requestInvoker;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<CohostOnAutoInsertEndpoint>();
+    private readonly IRazorDocumentMappingService _razorDocumentMappingService = razorDocumentMappingService;
+
+    protected override bool MutatesSolutionState => false;
+
+    protected override bool RequiresLSPSolution => true;
+
+    public Registration? GetRegistration(VSInternalClientCapabilities clientCapabilities, DocumentFilter[] filter, RazorCohostRequestContext requestContext)
+    {
+        if (clientCapabilities.SupportsVisualStudioExtensions)
+        {
+            // TODO: Add overload that doesn't use solution
+            //_remoteServiceProvider.TryInvokeAsync()
```

Comment:
Okay, I can see why `IRemoteAutoInsertService` has `TriggerCharacters` now, but as mentioned, I don't think its worth going to OOP for this info.

Perhaps `IOnAutoInsertProvider` can inherit from a new `IOnAutoInsertTriggerCharacterProvider` which has just the `TriggerCharacter` property, and this class can MEF import them?

Reply:
Didn't think it was a big deal since we only call it once, at the same time I guess it could slow down init so maybe the overhead is justified. I'll do that, let me know if that's not what you had in mind.

Reply:
What you have now is what I had in mind. It's a little clunky, I'll admit, but for accessing what is essentially a static list of `char`s I think it makes sense. Also its not strictly called once, its every time the language server starts, which is every time a solution is opened.